### PR TITLE
`urbit play`: replay event log subcommand

### LINKED ISSUE #157: Restore event log replay functionality.
> [Pull] the event log replay functionality out of v1.14 of Vere with the intention of it landing in v1.19.

> 1.14 was reverted due to some kind of corruption error but the error was from a different part of the release so the idea here is to extract the replay improvements from 1.14 and make a PR that just contains those

### PR DESCRIPTION
**Pull request by sigilante**
Resolves #157; see PR #158 for discussion.

Playback is normal speed and Ctrl+Z works to exit without hanging.  This may have inadvertently clobbered some of your replay stuff @mcevoypeter.

After `urbit play` succeeds, it boots into a prompt—not sure if this is desirable behavior or not.

### CODE REVIEW

File: pkg/noun/vortex.c [hunk 1]
```diff
@@ -384,4 +449,3 @@ u3v_rewrite_compact()
   arv_u->now = u3a_rewritten_noun(arv_u->now);
   arv_u->yot = u3a_rewritten_noun(arv_u->yot);
 }
-
```
Comment (joemfb) [inline pkg/noun/vortex.c:387]: This blank line at the end of the file should be restored.

File: pkg/vere/db/lmdb.h [hunk 2]
```diff
@@ -64,4 +75,22 @@
...
+    /* u3_lmdb_walk_next(): synchronously read next event from iterator.
+    */
+      c3_o
+      u3_lmdb_walk_next(u3_lmdb_walk* itr_u, size_t* len_i, void** buf_v);
+
+    /* u3_lmdb_walk_done(): close iterator.
+    */
+      void
+      u3_lmdb_walk_done(u3_lmdb_walk* itr_u);
+
```
Comment (joemfb) [inline pkg/vere/db/lmdb.h:95]: These `lmdb.(c|h)` changes were also copied into #165, which was just merged.

File: pkg/vere/main.c [hunk 3]
```diff
@@ -1714,55 +1727,155 @@ _cw_pack(c3_i argc, c3_c* argv[])
...
+    fprintf(stderr, "invalid command\r\n");
     exit(1);
   }
 
-  u3_Host.play_o = c3y;
+  //  XX handle SIGTSTP so that the lockfile is not orphaned?
+  //
+  u3_disk* log_u = _cw_disk_init(u3_Host.dir_c); // XX s/b try_aquire lock
+
+#if !defined(U3_OS_mingw)
```
Comment (joemfb) [inline pkg/vere/main.c:1827]: This `#if` can be removed.

File: pkg/vere/mars.c [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+  if ( u3C.wag_w & u3o_trace ) {
+    if ( u3_Host.tra_u.con_w == 0  && u3_Host.tra_u.fun_w == 0 ) {
+      u3t_trace_open(dir_c);
+    }
+    else if ( u3_Host.tra_u.con_w >= 100000 ) {
+      u3t_trace_close();
+      u3t_trace_open(dir_c);
+    }
+  }
+}
```
Comment (joemfb) [inline pkg/vere/mars.c:27]: The trace module was refactored upstream, so this function should change to match `_serf_step_trace()` (the only difference being that the pier path is passed in).

File: pkg/vere/mars.c [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+    return;
+  }
+  else {
+    eve_d = c3_min(eve_d, log_u->dun_d);
+  }
+
+  if ( !mar_u->dun_d ) {
+    c3_w lif_w;
+
+    if ( c3n == u3_disk_read_meta(log_u, 0, 0, &lif_w) ) {
```
Comment (joemfb) [inline pkg/vere/mars.c:171]: This signature also has been changed upstream.

File: pkg/vere/vere.h [hunk 6]
```diff
@@ -321,6 +330,7 @@
         c3_o       pep_o;                   //  prep for upgrade
         c3_o       play_o;                  //  replay event log and exit
         c3_i       xit_i;                   //  exit code for shutdown
+        u3_trac    tra_u;                   //  tracing information
```
Comment (joemfb) [inline pkg/vere/vere.h:333]: This struct member and the declaration can be (re)removed, as the tracing module has been refactored.

**Review by joemfb — CHANGES_REQUESTED** [review]
Thanks for restaging this, I'm really glad you got it working.

I've noted some minor things below. Also, please squash these commits down to one, and then merge `develop` back on top.

**Review by joemfb — CHANGES_REQUESTED** [review]
I forgot to mention: Please change the `_cw_play()` call site in `_cw_utils()` to return `1` (done) instead of `2` (continue).

File: pkg/vere/mars.c [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+    }
+  }
+}
+
+/* _mars_poke_play(): replay an event.
+*/
+static u3_weak
+_mars_poke_play(u3_mars* mar_u, c3_d eve_d, u3_noun job)
+{
+  fprintf(stderr, "play (%" PRIu64 ")\r\n", eve_d);
```
Comment (joemfb) [inline pkg/vere/mars.c:34]: Also, please remove this printf.

**Review by joemfb — CHANGES_REQUESTED** [review]

**Comment by sigilante** [comment]
Ah, that's why it was going to the Dojo prompt (1 v 2).

**Comment by sigilante** [comment]
Well, `git squash` and rebase looks like a royal mess here.

**Comment by sigilante** [comment]
@matthew-levan try this version now.

@belisarius222 the commit history is a mess here now so I may need you to show me how to rebase this against `develop` with a squash.

**Comment by matthew-levan** [comment]
@sigilante still hanging for me, on `linux-x86_64` and `macos-aarch64`, unfortunately. My test steps are: boot fresh fakezod, `rm zod/.urb/chk`, `./urbit play zod`.

**Comment by zalberico** [comment]
Can we fix the off-by-one error Peter fixed in https://github.com/urbit/vere/pull/192 in this one too?

File: pkg/vere/main.c (hunk 3 above)
Comment (joemfb) [reply pkg/vere/main.c]: Sorry for the confusion, I just meant for you to removed the condition, not the code inside of it. That code (setting up the ctrl-z handler) needed to be excluded on windows (which is no longer supported), so it should just be included unconditionally.

**Review by joemfb — CHANGES_REQUESTED** [review]
Sorry, my comment (https://github.com/urbit/vere/issues/157#issuecomment-1425069566) could've been more clear.

There are a some problems with your fixes in c48cf9807e8673b05e520a97d7f443b446598a83:

- the ctrl-z handler needs to be restored
- the changes to `serf.c` were desirable (`_serf_prof()` was moved to `u3a_prof()`, so should not be restored)
- the usage string that you added back from #192 doesn't not correspond to these arguments

Also, it looks like you rebased or cherry-picked `develop` onto this branch. There are conflict markers in several files, and many more commits than should be present in this PR view. The simplest thing is probably to force push this branch back to c48cf9807e8673b05e520a97d7f443b446598a83 and clean it up, then simply merge.

**Comment by joemfb** [comment]
@zalberico this PR adds an entirely different replay code path, the only area of overlap is in the interface (the play subcommand itself).

**Comment by joemfb** [comment]
@matthew-levan can you say any more about your testing? I've built this branch on `macos-aarch64` and replayed without issue.

**Comment by sigilante** [comment]
Current status:  there is some ongoing discussion in #157 about a discrepancy in playback speed dependent on the pill which boots Arvo.

That rather leads me to believe that what we're seeing with `urbit play` is a consequence not a cause, and we have something else going on that's simply manifesting here.

**Pull request closed without merging**